aten/src/ATen/SparseTensorImpl.h

```diff
@@ -8,27 +8,38 @@ namespace at {
 struct SparseTensorImpl : public TensorImpl {
   // Stored in COO format, indices + values.
 
+#ifndef USE_TH_SIZE_ZERO_DIM
   // Ideal INVARIANTS:
+#else
+  // INVARIANTS:
+#endif
   // _sparseDims: range [0, len(shape)]; _sparseDims + _denseDims = len(shape)
   // _denseDims : range [0, len(shape)]; _sparseDims + _denseDims = len(shape)
   // _indices.shape: dimensionality: 2,  shape: (_sparseDims, nnz)
   // _values.shape:  dimensionality: 1 + _denseDims.  shape: (nnz, shape[_sparseDims:])
 
+#ifndef USE_TH_SIZE_ZERO_DIM
   // Actual INVARIANT differences:
   // 1) _sparseDims: range [1, len(shape)] (i.e. we don't allow 0 sparse dimensions)
   // 2) when nnz = 0, there is strange behavior because we lack 0-dimensional sparse tensors.  Namely:
   //    dimensionality == 0, _sparseDims == 0, _denseDims == 0, _indices.shape == {0}, _values.shape == {0}
   // 3) For both _indices.shape and _values.shape, the nnz dimension may be larger than nnz
   // 4) For _values.shape, the non-nnz dimensions may be smaller than the corresponding dimension size, e.g.
   //    a shape (2,3) sparse tensor with _sparseDims == 1, may have _values.shape: (nnz, <=2, <=3).
+#endif
 
 
   // The true size of the sparse tensor (e.g., if you called to_dense()
   // on it).  When THTensor merges into TensorImpl, this field
   // should move to the parent class.
   std::vector<int64_t> size_;
 
+#ifndef USE_TH_SIZE_ZERO_DIM
   // The number of non-zero elements.
+#else
+  // The number of non-zero elements, which is guaranteed to match the
```

Comment:
this comment isn't that clear; I think you mean something like "match the corresponding(nnz) dimensions of indices/values.".

Reply:
Fixed